DI registration

Pliner opened the pull request:
Cherry picks from #1108 due to my "rebase" mistake.

----------------------------------------------------

Fixes https://github.com/EasyNetQ/EasyNetQ/issues/1042. This PR demonstrates the idea outlined in https://github.com/EasyNetQ/EasyNetQ/issues/1042. No breaking changes, only new APIs so targeted on master branch. In general it works, but I ran into the problem that not all DI containers support the concept of overriding registrations.
Now tests are falling for these containers:

Autofac - it is unlikely that anything can be done to full support replace semantics
SimpleInjector - [supports](https://simpleinjector.readthedocs.io/en/latest/howto.html#override-existing-registrations) replace semantics but has a specific API, which is not very convenient to use
Castle.Windsor - it seems to work, but it's not clear yet
MSDI container works great. Regarding https://github.com/EasyNetQ/EasyNetQ/issues/1042#issuecomment-621172418 comment from @zidad : in principle, I agree that we can stop supporting the rest of the containers and keep only MSDI. At least that's one what I'm working with 😄 . If I can't get the rest of the containers to work, then I will increasingly lean towards this option.

In any case, if we target this fix for the next major release (develop branch), then I can greatly simplify it by allowing a few backward incompatible changes. But I would not want to do that (at least for now), because now I only need a fix for duplicating registrations.

================= [UPDATES] ==================

IServiceRegister interface redesigned to support Add/Replace/TryAdd semantics: 3 Registed methods (with bool replace = true arg) + 3 TryRegister with mode` arg). All generic methods were moved into extensions.
Updated LightInject package + code unification with DefaultServiceContainer/LightInjectAdapter. One source file is used now for both ENQ internal implementation and EasyNetQ.DI.LightInject package.
Fix for InterceptorRegistrator to allow collection registration.

Pliner commented:
@sungam3r Could you please verify that all relevant commits are here?

Pliner commented on Source/EasyNetQ.DI.Tests/ContainerAdapterTests.cs:
```diff
@@ -39,92 +24,396 @@ public void Should_last_registration_win(ResolverFactory resolverFactory)
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_first_registration_win_type(string name, ResolverFactory resolverFactory)
+    {
+        //TODO: failed now
```
This kind of TODOs bothers me a lot: one implementation supports  some behaviour, another does not.

Pliner replied:
I'm not inclined to release PR in this state unless somebody reassures me that it is fine.

sungam3r replied:
>  one implementation supports some behaviour, another does not

Yep. This is reality.

Pliner commented:
Am I right that the things you are trying to fix in this PR are to following:
1. DI interface promises "first wins" semantic, but in fact this is not how it works. Currently, it seems to be "last wins".
2. There is an issue with configuration of injection of IEnumerable<SomeInterface>: it is not clear how to remove preconfigured implementations.

sungam3r commented:
Yes, I want to get rid of doubled registrations and fix behavior in such a way when default registrations are applied after user supplied ones, i.e. `TryRegister` instead of `Register`.

sungam3r commented:
> Could you please verify that all relevant commits are here?

I'll review entire diff tomorrow.

Pliner commented:
> Yes, I want to get rid of doubled registrations and fix behavior in such a way when default registrations are applied after user supplied ones, i.e. `TryRegister` instead of `Register`.

Cool. 

Might it be better just to apply `TryRegister` in separate branch and do not use a feature of registering/resolving `IEnumerable<SomeService>`? The PR adds complexity and it seems to be not a goal for EasyNetQ to be a full featured facade DI.

Also, I've checked how many usages each adapter has. In my opinion we should discontinue support for SimpleInjector and StructuredMap.

Pliner commented:
https://github.com/EasyNetQ/EasyNetQ/pull/1391

sungam3r commented:
I'm fine to get rid of SimpleInjector/StructureMap. So only Autofac has pending issues. Also see https://github.com/EasyNetQ/EasyNetQ/pull/1108#issuecomment-1096895923 .

>  Could you please verify that all relevant commits are here?

Done.

And I want to note again - event existing issues with Autofac do not prevent the use of ENQ as it is used now. Everything that was impossible to do (due to concrete DI container design) before this PR is also impossible to do after this PR.

Pliner commented:
As an intermediate version I suppose we could:
1. Include only part of @sungam3r PR, but do not include any of IEnumerable features(Register without replace and TryRegister with mode).
2. As addition, I've added proper(at least from my point of view) Autofac TryRegister implementation.
3. I've replaced usages of Shouldly with FluentAssertions.
4. I've slightly changed how interceptors work.

I hope that it should fix the issue with multiple registrations and there should be not so many objections about this current version.

sungam3r commented:
I'll review tomorrow morning.

sungam3r commented on Source/EasyNetQ/DI/IServiceRegister.cs:
```diff
@@ -3,50 +3,39 @@
     /// <summary>
-    /// Register a service. Note that the first registration wins. All subsequent registrations
-    /// will be ignored.
+    /// Registers(or replaces) the service of type <paramref name="serviceType"/> with the <paramref name="implementationType"/>
```
```suggestion
    /// Registers (or replaces if already registered) the service of type <paramref name="serviceType"/> with the <paramref name="implementationType"/>
```

sungam3r commented on Source/EasyNetQ/DI/IServiceRegister.cs:
```diff
@@ -3,50 +3,39 @@
     /// <summary>
-    /// Registers the <paramref name="serviceType"/> with the <paramref name="implementingType"/>.
-    /// Note that the first registration wins. All subsequent registrations will be ignored.
+    /// Tries to register(if it is not registered) the service of type <paramref name="serviceType"/> with the <paramref name="implementationType"/>
```
```suggestion
    /// Tries to register (if it is not registered) the service of type <paramref name="serviceType"/> with the <paramref name="implementationType"/>
```

sungam3r commented on Source/EasyNetQ.DI.Tests/ContainerAdapterTests.cs:
```diff
@@ -35,92 +24,229 @@ public void Should_last_registration_win(ResolverFactory resolverFactory)
+    [ClassData(typeof(ContainerAdaptersData))]
+    public void Should_resolve_single_registration(string name, ResolverFactory resolverFactory)
+    {
+        if (name == "Autofac")
```
Still here? You said thst you fixed Autofac registrations.

Pliner replied:
Yep. 

It is the only thing left(I've fixed TryRegister methods for Autofac, because it supports conditional registration). There is only one way at the moment to resolve this: to copy container builder each time when Register is called.

sungam3r replied:
OK. May be post "known issue" about such Autofac design?
